backend/src/libs/plugins/authorization/authorization.plugin.ts

```diff
@@ -0,0 +1,84 @@
+import fp from 'fastify-plugin';
+import { type JWTPayload } from 'jose';
+
+import {
+  HttpCode,
+  HttpError,
+  type HttpMethod,
+} from '~/libs/packages/http/http.js';
+import { type WhiteRoute } from '~/libs/packages/server-application/libs/types/types.js';
+import { token } from '~/libs/packages/token/token.js';
+import { type AuthService } from '~/packages/auth/auth.service';
+import { type UserService } from '~/packages/users/user.service';
+
+type Options = {
+  routesWhiteList: WhiteRoute[];
+  services: {
+    userService: UserService;
+    authService: AuthService;
+  };
+};
+
+interface IPayload extends JWTPayload {
+  id: number;
+}
+
+const errorMessage =
+  'You do not have the necessary authorization to access this resource. Please log in.';
+
+const authorization = fp(
+  (fastify, { routesWhiteList, services }: Options, done) => {
+    fastify.decorateRequest('user', null);
+
+    fastify.addHook('onRequest', async (request) => {
+      const isWhiteRoute = routesWhiteList.some((whiteRoute) => {
+        const isWhitePath = whiteRoute.routerPath === request.routerPath;
+        const isAllowedMethod = whiteRoute.methods.includes(
+          request.method as HttpMethod,
+        );
+        return isWhitePath && isAllowedMethod;
+      });
+
+      if (isWhiteRoute) {
+        return;
+      }
+
+      const authorizationHeader = request.headers.authorization;
+
+      if (!authorizationHeader) {
+        throw new HttpError({
+          status: HttpCode.UNAUTHORIZED,
+          message: errorMessage,
+        });
+      }
+
+      const [, requestToken] = authorizationHeader.split(' ');
+
+      if (!requestToken) {
+        throw new HttpError({
+          status: HttpCode.UNAUTHORIZED,
+          message: errorMessage,
+        });
+      }
+
+      const { userService } = services;
+      const { payload } = await token.verifyToken(requestToken);
+      const id = payload.id as IPayload['id'];
+
```

Comment:
And also add not empty `userId` assertion

Reply:
Was not resolved

Reply:
I am little confused. What do you mean?
In the next code userId has type number without null or underfine, so why do i need assertion?
 
` const { userId } = await token.verifyToken<{ userId: number }>(
        requestToken,
      );

      const authorizedUser = await userService.find(userId);`

Reply:
Because the fact that TS (by our typing) shows us that type is something that we declared does no mean that the value coming from outside would look like we declared - and as we know that from the outside somebody may sign token without `userId` in it - we may assert this value existence  ++ change generic type to more actual to reality 🤔 (`{ userId?: number }`)

Reply:
So do you mean to add if condition for userId?

Reply:
Yes